lucene/core/src/java/org/apache/lucene/index/EagerCheapMergePolicy.java

```diff
@@ -0,0 +1,202 @@
+
+  private static final String INFO_STREAM_CATEGORY = "CHEAP_MP";
+
+  private long cheapMergeThresholdBytes = 2 * 1024 * 1024; // 2MB
+  private int cheapMinMergeAtOnce = 3;
+  private double cheapMaxSizeRatio = 1.0;
+  private int cheapLimitConcurrentMergeSegments = 0; // thus only when no merges in-progress
+
+  /** Sole constructor. */
+  public EagerCheapMergePolicy() {
+  }
+
+  /** Returns {@link #setCheapMergeThresholdMB(double)}. */
+  public long getCheapMergeThresholdBytes() {
+    return cheapMergeThresholdBytes;
+  }
+
+  /** Limit cheap merges to those less than this size in megabytes. */
+  public EagerCheapMergePolicy setCheapMergeThresholdMB(double v) {
+    v *= 1024 * 1024;
+    cheapMergeThresholdBytes = (v > Long.MAX_VALUE) ? Long.MAX_VALUE : (long) v;
+    return this;
+  }
+
+  /** Returns {@link #setCheapMinMergeAtOnce(int)}. */
+  public int getCheapMinMergeAtOnce() {
+    return cheapMinMergeAtOnce;
+  }
+
+  /** Must have at least this many cheap segments to suggest merging them. */
+  public EagerCheapMergePolicy setCheapMinMergeAtOnce(int cheapMinMergeAtOnce) {
+    this.cheapMinMergeAtOnce = cheapMinMergeAtOnce;
+    return this;
+  }
+
+  /** Returns {@link #setCheapMaxSizeRatio(double)}. */
+  public double getCheapMaxSizeRatio() {
+    return cheapMaxSizeRatio;
+  }
+
+  /**
+   * Limit the maximum ratio of size between a candidate segment (that is deemed cheap) and the sum
+   * of all smaller segments. Example: If it's 2.0, a 20KB segment will merge with a 10KB
+   * segment (or say two 5KB segments) but will not with a 9KB segment.
+   * The limit should be tuned if {@link #setCheapMinMergeAtOnce(int)} is changed. It should be more
+   * than 1/(min-1).
+   */
+  public EagerCheapMergePolicy setCheapMaxSizeRatio(double cheapMaxSizeRatio) {
+    if (cheapMaxSizeRatio <= 0.0) {
+      throw new IllegalArgumentException("must be > 0");
+    }
+    this.cheapMaxSizeRatio = cheapMaxSizeRatio;
+    return this;
+  }
+
+  /** Returns {@link #setCheapLimitConcurrentMergeSegments(int)}. */
+  public int getCheapLimitConcurrentMergeSegments() {
+    return cheapLimitConcurrentMergeSegments;
+  }
+
+  /**
+   * Only find a cheap merge when the number of existing merging segments is less than or equal to
+   * this number.  In all likelihood, those segments are being merged concurrently. By default this
+   * setting is 0, thus won't find cheap merges during concurrent merging. If a higher number is
+   * used (e.g. 10) then there is a likely chance {@link org.apache.lucene.index.ConcurrentMergeScheduler}
+   * will intentionally stall the current thread if it's thresholds have been reached.  This may be
+   * okay if there are multiple merge threads (thus less likely to reach such thresholds) or if it's
+   * deemed more important to do cheap merges at the expense of slowing index throughput further.
+   */
+  public EagerCheapMergePolicy setCheapLimitConcurrentMergeSegments(int cheapLimitConcurrentMergeSegments) {
```

Comment:
I think this is somewhat dangerous .... when a large merge kicks off, and runs for minutes maybe, it means we would suppress these "cheap merges on refresh/commit" while it is running?

Reply:
Yes but why is that dangerous?  The intent of this was to protect from excessive merging / i/o.

Reply:
Well, dangerous because it leads to merge starvation.

Those tiny/fast merges really need to run at all times.  If you stop them because a huge merge is running it can lead to an explosion of small segments in the index, GC problems, running out of file descriptors, etc.

Reply:
Actually, I was confused here -- since this merge policy is only picking (cheap) merges during commit, it is not as dangerous as I thought.  I had thought it was doing this in general (for natural and forced merges too).